### ISSUE TITLE
Endpoint timeout not respected when connecting

### ISSUE THREAD

**Issue by paddor** [socketry/async-http#184]
Given a host that is unreachable:
```
$ traceroute -m10 192.168.99.1
traceroute to 192.168.99.1 (192.168.99.1), 10 hops max, 60 byte packets
 1  * * *
 2  * * *
 3  * * *
 4  * * *
 5  * * *
 6  * * *
 7  * * *
 8  * * *
 9  * * *
10  * * *
```

And Gemfile:
```ruby
source "https://rubygems.org"

gem 'async-http'
```

The script with a `Async::HTTP::Endpoint#timeout` set:

```ruby
require 'async/http'

HOST             = '192.168.99.1'
CONNECT_ENDPOINT = Async::HTTP::Endpoint.for 'http', HOST, '/idp/connect/token', timeout: 3

CONNECT_DATA     = {
  grant_type: 'password',
  username:   'user',
  password:   'foobar',
}

Sync do
  headers = {
    'accept'       => 'application/json',
    'Content-Type' => 'application/x-www-form-urlencoded',
  }
  body = URI.encode_www_form CONNECT_DATA

  client = Async::HTTP::Client.new CONNECT_ENDPOINT
  response = client.post CONNECT_ENDPOINT, headers, body
end
```

It does not timeout:

```
$ time env CONSOLE_LEVEL=debug bundle exec async_connect_timeout.rb
  0.0s    debug: Async::Pool::Controller [oid=0x7d0] [ec=0x7e4] [pid=2382442] [2024-10-01 13:07:23 +0200]
               | Acquiring concurrency guard...
               | {
               |   "blocking": false
               | }
  0.0s    debug: Async::Pool::Controller [oid=0x7d0] [ec=0x7e4] [pid=2382442] [2024-10-01 13:07:23 +0200]
               | Acquired concurrency guard.
  0.0s    debug: Async::Pool::Controller [oid=0x7d0] [ec=0x7e4] [pid=2382442] [2024-10-01 13:07:23 +0200]
               | No available resources, allocating new one...
  0.0s    debug: Async::HTTP::Client [oid=0x80c] [ec=0x7e4] [pid=2382442] [2024-10-01 13:07:23 +0200]
               | Making connection to #<Async::HTTP::Endpoint http://192.168.99.1/idp/connect/token {:timeout=>3}>
^C10.59s    debug: Async::Pool::Controller: Async::Pool::Controller Gardener [oid=0x7d0] [ec=0x820] [pid=2382442] [2024-10-01 13:07:34 +0200]
               | Draining pool...
               | {
               |   "size": 0
               | }
/home/user/.gem/ruby/3.3.4/gems/async-2.17.0/lib/async/scheduler.rb:317:in `select': Interrupt
        from /home/user/.gem/ruby/3.3.4/gems/async-2.17.0/lib/async/scheduler.rb:317:in `run_once!'
        from /home/user/.gem/ruby/3.3.4/gems/async-2.17.0/lib/async/scheduler.rb:353:in `run_once'
        from /home/user/.gem/ruby/3.3.4/gems/async-2.17.0/lib/async/scheduler.rb:408:in `block in run'
        from /home/user/.gem/ruby/3.3.4/gems/async-2.17.0/lib/async/scheduler.rb:386:in `block in run_loop'
        from /home/user/.gem/ruby/3.3.4/gems/async-2.17.0/lib/async/scheduler.rb:383:in `handle_interrupt'
        from /home/user/.gem/ruby/3.3.4/gems/async-2.17.0/lib/async/scheduler.rb:383:in `run_loop'
        from /home/user/.gem/ruby/3.3.4/gems/async-2.17.0/lib/async/scheduler.rb:407:in `run'
        from /home/user/.gem/ruby/3.3.4/gems/async-2.17.0/lib/kernel/sync.rb:28:in `Sync'
        from async_connect_timeout.rb:15:in `<top (required)>'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/3.3.0/bundler/cli/exec.rb:58:in `load'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/3.3.0/bundler/cli/exec.rb:58:in `kernel_load'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/3.3.0/bundler/cli/exec.rb:23:in `run'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/3.3.0/bundler/cli.rb:455:in `exec'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/3.3.0/bundler/vendor/thor/lib/thor/command.rb:28:in `run'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/3.3.0/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/3.3.0/bundler/vendor/thor/lib/thor.rb:527:in `dispatch'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/3.3.0/bundler/cli.rb:35:in `dispatch'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/3.3.0/bundler/vendor/thor/lib/thor/base.rb:584:in `start'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/3.3.0/bundler/cli.rb:29:in `start'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/gems/3.3.0/gems/bundler-2.5.11/exe/bundle:28:in `block in <top (required)>'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/3.3.0/bundler/friendly_errors.rb:117:in `with_friendly_errors'
        from /home/user/.rubies/ruby-3.3.4/lib/ruby/gems/3.3.0/gems/bundler-2.5.11/exe/bundle:20:in `<top (required)>'
        from /home/user/.gem/ruby/3.3.4/bin/bundle:25:in `load'
        from /home/user/.gem/ruby/3.3.4/bin/bundle:25:in `<main>'

________________________________________________________
Executed in   10.96 secs      fish           external
   usr time  307.06 millis    0.00 micros  307.06 millis
   sys time   51.88 millis  864.00 micros   51.01 millis
```

Is there a way to achieve a connect timeout without wrapping it in `Async::Task#with_timeout`?

**Comment by ioquatix** [comment]
Yes, we should use the default timeout for connect too.

**Comment by ioquatix** [comment]
I checked, and the code in `io-endpoint` is correctly assigning the timeout, but `Socket#connect` does not appear to be respecting it. It looks like a bug in CRuby, so I'll need to investigate further.

**Comment by paddor** [comment]
That sounds great. Thank you.

**Comment by ioquatix** [comment]
I've merged this into CRuby. It will be fixed when running on Ruby 3.4+. It's not an issue we can address in `async-http` unfortunately.
